heracles/dices/shrinkage.py

```diff
@@ -103,76 +94,75 @@ def shrinkage_factor(cls1, target):
     return lambda_star
 
 
-def gaussian_covariance(Cls):
-    """
-    Computes Gaussian estimate of the target matrix.
-    input:
-        Cls: power spectra
-    returns:
-        T: target matrix
-    """
-    # Add bias to Cls
-    b = bias(Cls)
-    Cls = add_to_Cls(Cls, b)
-    # Separate Cls into Cls
-    _Cls = _fields2components(Cls)
-    # Compute Gaussian covariance
+def gaussian_covariance(cls):
+    b = bias(cls)
+    cls = add_to_Cls(cls, b)
     cov = {}
-    for key1, key2 in itertools.combinations_with_replacement(Cls, 2):
-        # covariance key
+    for key1, key2 in itertools.combinations_with_replacement(cls.keys(), 2):
         a1, b1, i1, j1 = key1
         a2, b2, i2, j2 = key2
         covkey = (a1, b1, a2, b2, i1, j1, i2, j2)
         # get reference results
-        cl1 = Cls[key1]
-        cl2 = Cls[key2]
+        cl1 = cls[key1]
+        cl2 = cls[key2]
         sa1, sb1 = cl1.spin
         sa2, sb2 = cl2.spin
-        # get components
-        _a1, idx1 = _split_key(a1, sa1, pos=0)
-        _b1, idx2 = _split_key(b1, sb1, pos=0)
-        _a2, idx3 = _split_key(a2, sa2, pos=0)
-        _b2, idx4 = _split_key(b2, sb2, pos=0)
+        # dof spins
+        dof_a1 = 1 if sa1 == 0 else 2
+        dof_b1 = 1 if sb1 == 0 else 2
+        dof_a2 = 1 if sa2 == 0 else 2
+        dof_b2 = 1 if sb2 == 0 else 2
         # get attributes of result
-        ell1 = get_result_array(cl1, "ell")
-        ell2 = get_result_array(cl2, "ell")
-        ell = ell1 + ell2
-        r = np.zeros(
-            (len(idx1), len(idx2), len(idx3), len(idx4), len(ell1[0]), len(ell2[0]))
-        )
-        # get covariance
-        for k, idx in zip(
-            itertools.product(_a1, _b1, _a2, _b2),
-            itertools.product(idx1, idx2, idx3, idx4),
-        ):
-            __a1, __b1, __a2, __b2 = k
-            _key = (__a1, __b1, __a2, __b2, i1, j1, i2, j2)
-            _cov = _gaussian_covariance(_Cls, _key)
-            ix1, ix2, ix3, ix4 = idx
-            r[ix1, ix2, ix3, ix4, :, :] = np.diag(_cov)
-        # Remove the extra dimensions
+        ell1 = get_result_array(cl1, "ell")[0]
+        ell2 = get_result_array(cl2, "ell")[0]
+        # keys for cov
+        _key1 = (a1, a2, i1, i2)
+        _key2 = (b1, b2, j1, j2)
+        _key3 = (a1, b2, i1, j2)
+        _key4 = (b1, a2, j1, i2)
+        _cl1 = get_cl(_key1, cls)
+        _cl2 = get_cl(_key2, cls)
+        _cl3 = get_cl(_key3, cls)
+        _cl4 = get_cl(_key4, cls)
```

Comment:
Shorter is clearer here I think.


```suggestion
        _cl1 = get_cl((a1, a2, i1, i2), cls)
        _cl2 = get_cl((b1, b2, j1, j2), cls)
        _cl3 = get_cl((a1, b2, i1, j2), cls)
        _cl4 = get_cl((b1, a2, j1, i2), cls)
```

[Although looking at it now, I think the more natural signature would have been `get_cl(cls, key)`, similar to `__getitem__(self, key)`.]